File: flink-udf/src/main/java/com/longyun/flink/java/udf/JsonRow.java
package com.longyun.flink.java.udf;

import org.apache.flink.table.api.functions.TableFunction;
import org.apache.flink.table.api.types.DataType;
import org.apache.flink.table.api.types.DataTypes;
import org.apache.flink.table.api.types.RowType;
import org.apache.flink.table.runtime.functions.utils.JsonUtils;
import org.apache.flink.types.Row;

import java.sql.Timestamp;
import java.sql.Date;
import java.sql.Time;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;


public class JsonRow extends TableFunction<Row> {

    private Map<String, DataType> fieldTypeMap;


    public JsonRow(Map<String, DataType> fieldTypeMap){
        this.fieldTypeMap = fieldTypeMap;
    }

    public void eval(String jsonStr, String... paths) {
        Row row = new Row(paths.length);
        try {
            for (int i = 0; i < paths.length; i++) {
                String value = JsonUtils.getInstance().getJsonObject(jsonStr, "$."+paths[i]);
                DataType typeName = fieldTypeMap.getOrDefault(paths[i], DataTypes.STRING);
                if(typeName == DataTypes.FLOAT){
                    row.setField(i, Float.parseFloat(value));
                }else if(typeName == DataTypes.DOUBLE){
                    row.setField(i, Double.parseDouble(value));
                }else if(typeName == DataTypes.INT){
                    row.setField(i, Integer.parseInt(value));
                }else if(typeName == DataTypes.LONG){
                    row.setField(i, Long.parseLong(value));
                }else if(typeName == DataTypes.TIMESTAMP){
                    Long v = Long.parseLong(value);
                    row.setField(i, new Timestamp(v));
                }else if(typeName == DataTypes.TIME){
                    Long v = Long.parseLong(value);
                    row.setField(i, new Time(v));
                }else if(typeName == DataTypes.DATE){
                    Long v = Long.parseLong(value);
                    row.setField(i, new Date(v));
                }else {
                    row.setField(i, value);
                }
            }

            collect(row);
        }catch (RuntimeException e){
            e.printStackTrace();
        }
    }


    @Override
    public DataType getResultType(Object[] arguments, Class[] argTypes) {
        List<DataType> types = new ArrayList<>();
        for (int i = 0; i < arguments.length; i++) {
            if(null == arguments[i]) continue;
            types.add(fieldTypeMap.getOrDefault(arguments[i].toString(), DataTypes.STRING));
        }

        DataType[] typeArray = new DataType[types.size()];
        types.toArray(typeArray);
        return new RowType(typeArray);
    }
}

File: flink-udf/src/main/scala/com/longyun/flink/udf/scala/JsonRow.scala
package com.longyun.flink.udf.scala

import java.sql.{Date, Timestamp, Time}

import org.apache.flink.table.api.functions.TableFunction
import org.apache.flink.table.api.types.{DataType, DataTypes, RowType}
import org.apache.flink.table.runtime.functions.utils.JsonUtils
import org.apache.flink.types.Row

import scala.annotation.varargs
import scala.collection.mutable.ArrayBuffer

class JsonRow(val fieldTypeMap:Map[String, DataType]) extends TableFunction[Row] {

  @varargs
  def eval(jsonStr: String, paths: String*): Unit = {
    println("jsonStr="+jsonStr)
    val row = new Row(paths.length)
    for (i <- 0 until paths.length) {
      val value = JsonUtils.getInstance.getJsonObject(jsonStr, "$."+paths(i))
      val typeName = fieldTypeMap.getOrElse(paths(i), DataTypes.STRING)
      typeName match {
        case DataTypes.FLOAT => row.setField(i, java.lang.Float.parseFloat(value))
        case DataTypes.INT => row.setField(i, java.lang.Integer.parseInt(value))
        case DataTypes.DOUBLE => row.setField(i, java.lang.Double.parseDouble(value))
        case DataTypes.LONG => row.setField(i, java.lang.Long.parseLong(value))
        case DataTypes.DATE => {
          val v = java.lang.Long.parseLong(value)
          row.setField(i, new Date(v))
        }
        case DataTypes.TIMESTAMP => {
          val v = java.lang.Long.parseLong(value)
          row.setField(i, new Timestamp(v))
        }
        case DataTypes.TIME => {
          val v = java.lang.Long.parseLong(value)
          row.setField(i, new Time(v))
        }
        case _ => row.setField(i, value)
      }
    }

    collect(row)
  }

  override def getResultType(arguments: Array[AnyRef], argTypes: Array[Class[_]]): DataType = {
    import util.control.Breaks._
    val fields = new ArrayBuffer[String]()
    val types = new ArrayBuffer[DataType]()
    for(i <- 0 until argTypes.length){
      breakable{
        if(arguments(i) == null) {
          break
        }
        fields += (arguments(i).toString)
        types += fieldTypeMap.getOrElse((arguments(i).toString), DataTypes.STRING)
      }
    }

    new RowType(types.toArray, fields.toArray)
  }
}